Introducing ULIDs

hackaugusto opened the pull request:
This PR ended being a bigger than what I wanted. Here is what it does:

- Remove the dependency on `lastrowid` [example](https://github.com/raiden-network/raiden/blob/0ae7627d8e484c78b35f37e3b2f5d660b5ca670d/raiden/storage/sqlite.py#L146), which [requires synchronization](https://github.com/raiden-network/raiden/blob/0ae7627d8e484c78b35f37e3b2f5d660b5ca670d/raiden/storage/sqlite.py#L89-L102) and [does not work with `executemany`](https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.lastrowid).
  - This removed some race conditions, which on the current code base were unlikely to be hit. ([Inserts that are not protected by the write lock](https://github.com/raiden-network/raiden/blob/0ae7627d8e484c78b35f37e3b2f5d660b5ca670d/raiden/storage/sqlite.py#L113-L118) which may result in the wrong `lastrowid` being used)
- Without `lastrowid` a new method to determine the IDs of the insert rows is necessary. This moves the ID generation from the database to the APP. This PR implements the IDs with ULIDs.
- ULIDs are used over UUID because Raiden requires total order for the entries [on restarts](https://github.com/raiden-network/raiden/blob/0ae7627d8e484c78b35f37e3b2f5d660b5ca670d/raiden/storage/wal.py#L19-L21), otherwise recovery is not deterministic.
- Because of the above our implementation requires a monotonic clock, and the wall time cannot be used. To handle this the `ULIDMonotonicFactory` was introduced, which deviates from the `ULID` spec by using `ns` resolution instead of `ms` (An alternative is to add a monotonic counter for the IDs generated on the same ms).
- Moved the `TIMESTAMP` generation to the database, since there is not many reasons to have Raiden providing it. The column was added to all WAL related tables. The timestamp was maintained because the ULID uses the monotonic clock which has no relation to wall time.
- I also did a review of SQLite's clustered IDs found some documented bugs that were not handled. Namely the `NOT NULL` constraint must always be added for non clustered indexes (everything but `INTEGER PRIMARY KEY` or tables with `WITHOUT ROWID`).

Review after: #4143

ulope commented:
Just FYI: On oldish linux kernels and certain hardware there is apparently a bug that can cause `CLOCK_MONOTONIC` to go backwards (that was recently an issue [Rust](https://github.com/rust-lang/rust/issues/37902) ran [into](https://github.com/longsleep/linux-pine64/issues/44)).

The suggestion seems to be to use `CLOCK_MONOTONIC_RAW` on linux, which is also [available in Python](https://docs.python.org/3/library/time.html#clock-id-constants)

hackaugusto commented:
Some additional notes about this PR:

- Removed the values `earliest` and `latest` from the database. With these values the query type was `Union[ULID, Literal["earliest"], Literal["latest"]]`. Now the type is simply `ULID`, `earliest` can be represented as the first value after `0`, and `latest` is the last before `2**256-1`.
- Consolidated the queries as ranges, instead of having `def get_x(self, start: ULID, end: Optional[ULID]):`, now the queries are done with `def get_x(self, range: Range):`, where `Range` is a close range `[from,to]`, and the object's constructor checks the values.
- Added proper generics typing to check the usage of `ULID`s, and to separate the raw database from the codec wrapper. 
- Added type annotation some of the sqlite tests, because the previous change broke some code and was only caught by the integration tests.

hackaugusto commented on raiden/blockchain_events_handler.py:
```diff
@@ -447,7 +447,9 @@ def handle_channel_batch_unlock(raiden: "RaidenService", event: Event):
             if state_change_record is not None:
-                canonical_identifier = state_change_record.data.balance_proof.canonical_identifier
+                canonical_identifier = (
+                    state_change_record.data.balance_proof.canonical_identifier  # type: ignore
```
I thought about introducing a new marker type with `balance_proof` in it. However this PR was big enough already.

hackaugusto commented on raiden/storage/sqlite.py:
```diff
@@ -1,33 +1,58 @@
+
+
+FIRST_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
+LAST_STATECHANGE_ULID = StateChangeID(ULID((2 ** 128 - 1).to_bytes(16, "big")))
```
These are the first and last possible values for the IDs, not the first and last value in the database.

konradkonrad replied:
```suggestion
LOW_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
HIGH_STATECHANGE_ULID = StateChangeID(ULID((2 ** 128 - 1).to_bytes(16, "big")))
```

hackaugusto commented on raiden/storage/sqlite.py:
```diff
@@ -51,21 +76,38 @@ class FilteredDBQuery(NamedTuple):
 
 
+class EventEncodedRecord(NamedTuple):
+    event_identifier: EventID
```
I did not consolidated the three types because their attributes are different.

konradkonrad replied:
I would still like to see the record primary identifier field names be in unison with the database layout: see https://github.com/raiden-network/raiden/issues/4078

hackaugusto commented on raiden/storage/sqlite.py:
```diff
@@ -180,22 +233,62 @@ def __init__(self, database_path: Path):
-        # Improve on this and find a better way to protect against this potential race
-        # condition.
         self.conn = conn
-        self.write_lock = threading.Lock()
```
This is one of the main reasons to have ULIDs, now we can insert entries in bulk and the IDs will be known.

hackaugusto commented on raiden/storage/sqlite.py:
```diff
@@ -477,40 +557,35 @@ def update_state_changes(self, state_changes_data: List[Tuple[str, int]]) -> Non
 
         cursor = self.conn.cursor()
 
-        if to_identifier == "latest":
```
Because `latest` is not supported anymore, this extra query is unnecessary.

hackaugusto commented on raiden/tests/utils/factories.py:
```diff
@@ -147,6 +147,10 @@ def make_uint64() -> int:
     return random.randint(0, UINT64_MAX)
 
 
+def make_payment_id() -> PaymentID:
```
These were not necessary for this PR, I did it while chasing a bug.

konradkonrad commented on raiden/storage/sqlite.py:
```diff
@@ -1,33 +1,58 @@
+            raise ValueError("last must be larger or equal to first")
+
+
+FIRST_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
```
```suggestion
LOW_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
```

konradkonrad commented on raiden/storage/sqlite.py:
```diff
@@ -1,33 +1,58 @@
+
+
+FIRST_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
+LAST_STATECHANGE_ULID = StateChangeID(ULID((2 ** 128 - 1).to_bytes(16, "big")))
```
```suggestion
HIGH_STATECHANGE_ULID = StateChangeID(ULID((2 ** 128 - 1).to_bytes(16, "big")))
```

konradkonrad commented on raiden/storage/sqlite.py:
```diff
@@ -1,33 +1,58 @@
+
+FIRST_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
+LAST_STATECHANGE_ULID = StateChangeID(ULID((2 ** 128 - 1).to_bytes(16, "big")))
+RANGE_ALL_STATE_CHANGES = Range(FIRST_STATECHANGE_ULID, LAST_STATECHANGE_ULID)
```
```suggestion
RANGE_ALL_STATE_CHANGES = Range(LOW_STATECHANGE_ULID, HIGH_STATECHANGE_ULID)
```

konradkonrad commented on raiden/storage/sqlite.py:
```diff
@@ -74,6 +116,14 @@ def assert_sqlite_version() -> bool:  # pragma: no unittest
     return True
 
 
+def adapt_ulid_identifier(ulid: ULID) -> bytes:
```
why the specialized naming?
```suggestion
def ulid_to_bytes_identifier(ulid: ULID) -> bytes:
```
or even just
```suggestion
def ulid_to_bytes(ulid: ULID) -> bytes:
```

I would argue that the `adapt_` / `convert_` pattern is mentally hard...

konradkonrad replied:
I see that these are nowhere to be used, so ignore...

konradkonrad commented on raiden/storage/sqlite.py:
```diff
@@ -729,23 +811,34 @@ def get_latest_state_change_by_data_field(
         """ Return all state changes filtered by a named field and value."""
 
-        state_change = self.database.get_latest_state_change_by_data_field(query)
+        state_change_encoded = self.database.get_latest_state_change_by_data_field(query)
```
just for some consistency, I would have loved to read
```
encoded_state_change...
```
here instead. But nvm.

konradkonrad commented on raiden/storage/utils.py:
```diff
@@ -1,3 +1,44 @@
+ULIDs [7]. ULIDs are necessary because the application must know the ID of
+elements inserted in bulk, and this cannot be know reliably if auto increment
+is used, even with synchronizaiton over `lastrowid`. Additionally random IDs
+are less likely to collide with is useful to detect programming errors. Typos
```
```suggestion
are less likely to collide, which is useful to detect programming errors. Typos
```

konradkonrad commented on raiden/storage/utils.py:
```diff
@@ -1,3 +1,44 @@
+prevent rewrites of the PK Index.
+
+Manifest typing [4] is used for the ULIDs instead PARSE_COLNAMES [5] because
+of the easy of convertion.
```
```suggestion
of the ease of conversion.
```

BTW: I really like this well written documentation!

konradkonrad commented on raiden/storage/sqlite.py:
```diff
@@ -661,46 +750,38 @@ def get_version(self) -> RaidenDBVersion:
-
-    def get_snapshot_closest_to_state_change(
-        self, state_change_identifier: Union[StateChangeID, str]
+    def get_snapshot_before_state_change(
```
I like the much clearer name here: "before" instead of "closest" :+1:

konradkonrad commented on raiden/storage/ulid.py:
```diff
@@ -0,0 +1,100 @@
+        assert len(self.identifier) == 16, "id_ must be 16 bytes long"
+
+    @property
+    def timestamp(self):
```
great, I was hoping for some "human readable" representation of ulids, for any sort of debugging...

konradkonrad commented on raiden/storage/ulid.py:
```diff
@@ -0,0 +1,100 @@
+
+
+class ULIDMonotonicFactory(Generic[ID]):
+    """Monotonic ULID factory that guarantees new ULIDs will not be decrease.
```
```suggestion
    """Monotonic ULID factory that guarantees new ULIDs will not decrease.
```

konradkonrad commented on raiden/tests/unit/test_sqlite.py:
```diff
@@ -155,7 +160,7 @@ def test_get_state_change_with_balance_proof():
     """ All state changes which contain a balance proof must be found by when
```
```suggestion
    """ All state changes which contain a balance proof must be found when
```

konradkonrad commented on setup.cfg:
```diff
@@ -50,3 +50,9 @@ warn_unused_ignores = True
+ignore_errors = False
+
+[mypy-raiden.tests.unit.test_sqlite]
+ignore_errors = False
```
:+1:

konradkonrad commented on raiden/storage/sqlite.py:
```diff
@@ -477,40 +557,35 @@ def update_state_changes(self, state_changes_data: List[Tuple[str, int]]) -> Non
+            "SELECT identifier, data "
+            "FROM state_changes "
+            "WHERE identifier "
+            "BETWEEN ? AND ? "
```
does the `BETWEEN` operator work bi-directionally? IOW are the queries:
```sql
[...]
BETWEEN LOW_STATECHANGE_ULID AND HIGH_STATECHANGE_ULID
[...]
```
and
```sql
[...]
BETWEEN HIGH_STATECHANGE_ULID AND LOW_STATECHANGE_ULID
[...]
```
equivalent? Otherwise I suggest a validation for the range above!

hackaugusto replied:
> does the BETWEEN operator work bi-directionally?

It does not:

> "x BETWEEN y AND z" is equivalent to "x>=y AND x<=z" 

https://www.sqlite.org/lang_expr.html#between

That is why I introduced the type `Range`, it cannot be constructed if the range is invalid:

https://github.com/raiden-network/raiden/pull/4156/files#diff-34fcdfdd66b47942cc02bf2a12fa4925R48
